File: src/core/guid.ts
import crypto from 'crypto';

// Lowercase letters and numbers only (36 chars) - easier to type than mixed case
const ALPHABET = '0123456789abcdefghijklmnopqrstuvwxyz';
const GUID_LENGTH = 8;
const DISPLAY_LENGTH_SMALL = 4;
const DISPLAY_LENGTH_MEDIUM = 5;
const DISPLAY_LENGTH_LARGE = 6;

/**
 * Generate a short GUID with a prefix (e.g., msg-xxxx).
 * Uses lowercase alphanumeric characters for easy typing.
 */
export function generateGuid(prefix: string): string {
  const normalizedPrefix = prefix.endsWith('-') ? prefix.slice(0, -1) : prefix;
  const bytes = crypto.randomBytes(GUID_LENGTH);
  let id = '';

  for (let i = 0; i < GUID_LENGTH; i++) {
    id += ALPHABET[bytes[i] % ALPHABET.length];
  }

  return `${normalizedPrefix}-${id}`;
}

export function getDisplayPrefixLength(messageCount: number): number {
  if (messageCount < 500) {
    return DISPLAY_LENGTH_SMALL;
  }
  if (messageCount < 1500) {
    return DISPLAY_LENGTH_MEDIUM;
  }
  return DISPLAY_LENGTH_LARGE;
}

export function getGuidPrefix(guid: string, length: number): string {
  const base = guid.startsWith('msg-') ? guid.slice(4) : guid;
  if (length <= 0) return '';
  return base.slice(0, Math.min(length, base.length));
}

File: internal/core/guid.go
package core

import (
	"crypto/rand"
	"fmt"
)

const (
	guidAlphabet        = "0123456789abcdefghijklmnopqrstuvwxyz"
	guidLength          = 8
	displayLengthSmall  = 4
	displayLengthMedium = 5
	displayLengthLarge  = 6
)

// GenerateGUID creates a short GUID with the provided prefix.
func GenerateGUID(prefix string) (string, error) {
	normalized := prefix
	if len(normalized) > 0 && normalized[len(normalized)-1] == '-' {
		normalized = normalized[:len(normalized)-1]
	}

	buf := make([]byte, guidLength)
	if _, err := rand.Read(buf); err != nil {
		return "", fmt.Errorf("generate guid: %w", err)
	}

	id := make([]byte, guidLength)
	for i := 0; i < guidLength; i++ {
		id[i] = guidAlphabet[int(buf[i])%len(guidAlphabet)]
	}

	return fmt.Sprintf("%s-%s", normalized, string(id)), nil
}

// GetDisplayPrefixLength returns the short GUID length for display.
func GetDisplayPrefixLength(messageCount int) int {
	if messageCount < 500 {
		return displayLengthSmall
	}
	if messageCount < 1500 {
		return displayLengthMedium
	}
	return displayLengthLarge
}

// GetGUIDPrefix extracts the shortened ID prefix used in UI.
func GetGUIDPrefix(guid string, length int) string {
	base := guid
	if len(base) >= 4 && base[:4] == "msg-" {
		base = base[4:]
	}
	if length <= 0 {
		return ""
	}
	if length > len(base) {
		length = len(base)
	}
	return base[:length]
}
